treewide: Nuke foreign_data_t

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

No longer used or needed.

The purpose of `foreign_data_t` was to avoid unsafe/undefined behavior when sharing a `record_batch_reader` between shards. This is no longer necessary since sharing an iobuf was made thread safe in https://github.com/redpanda-data/seastar/pull/149.

https://github.com/redpanda-data/redpanda/pull/24047 removed all uses of `foreign_data_t` but left behind the variant type in `record_batch_reader`. This PR removes all reference to `foreign_data_t` from the tree.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
treewide: Remove record_batch_reader::foreign_data_t

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

WillemKauf commented on src/v/model/record_batch_reader.h:
```diff
@@ -302,12 +259,12 @@ class record_batch_reader final {
-        // note that if the reader wasn't obtained through the cache, this can
-        // still return true even though the reader will ultimately not be
-        // cached
+        // note that if the reader wasn't obtained through the cache, this
```
nit: your IDE touched some comment formatting throughout the file, please checkout these chunks from the commit

oleiman replied:
ah yeah, good catch.

[naughty IDE](https://www.spacemacs.org/)

oleiman replied:
done

WillemKauf reviewed (dismissed):
woohoo

WillemKauf commented on src/v/model/record_batch_reader.h:
```diff
@@ -47,11 +47,7 @@ concept ReferenceBatchReaderConsumer = requires(
-        size_t index{0};
-    };
-    using storage_t = std::variant<data_t, foreign_data_t>;
+    using storage_t = data_t;
```
might be worth `sed` replacing and removing `storage_t` if possible, but I don't have too strong an opinion on that.

oleiman replied:
done

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b97b5893-b004-4724-8f3e-9e2c334ef298-->
<!--TEST RESULTS-->
<!--0c4aff6c-583f-4a5d-ac36-883b09a45982-->
<!--TEST RESULTS-->
<!--93833029-48c3-4047-8f1a-2d7da38208d6-->
<!--TEST RESULTS-->
<!--4dfa64b5-4aeb-4dd4-81e0-4dda2c61bcab-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66389">build#66389</a></summary>


| test_class                   | test_method       | test_arguments         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------|:------------------|:-----------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AccessControlListTestUpgrade | test_upgrade_sasl |                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66389#0196ff70-1f2e-4479-849a-161720dfd944 | FLAKY         | 20/21    | upstream reliability is '97.48858447488584'. current run reliability is '95.23809523809523'. drift is 2.25049 and the allowed drift is set to 50. The test should PASS |
| AuditLogTestOauth            | test_kafka_oauth  | {"authz_match": "acl"} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66389#0196ff70-1f2e-4479-849a-161720dfd944 | FLAKY         | 20/21    | upstream reliability is '97.47368421052632'. current run reliability is '95.23809523809523'. drift is 2.23559 and the allowed drift is set to 50. The test should PASS |
| DisablingPartitionsTest      | test_disable      |                        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66389#0196ff70-1f2f-45b6-a132-e0d85eb1a814 | FLAKY         | 14/21    | upstream reliability is '79.66386554621849'. current run reliability is '66.66666666666666'. drift is 12.9972 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66451">build#66451</a></summary>


| test_class                | test_method                             | test_arguments                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:----------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| OffsetForLeaderEpochTest  | test_offset_for_leader_epoch_transfer   |                                                                                                                                      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66451#01971326-55df-45fa-9e72-735f647471fc | FAIL          | 0/1      |                                                                                                                                                                         |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66451#01971326-55df-45fa-9e72-735f647471fc | FAIL          | 0/1      |                                                                                                                                                                         |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66451#01971326-55df-45fa-9e72-735f647471fc | FAIL          | 0/1      |                                                                                                                                                                         |
| RandomNodeOperationsTest  | test_node_operations                    | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": true, "with_chunked_compaction": false, "with_iceberg": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66451#01971326-55e2-4909-8853-026bd903bcb1 | FLAKY         | 16/21    | upstream reliability is '99.01960784313727'. current run reliability is '76.19047619047619'. drift is 22.82913 and the allowed drift is set to 50. The test should PASS |
| RecreateTopicMetadataTest | test_recreated_topic_metadata_are_valid | {"replication_factor": 3}                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66451#01971322-ed0f-45d5-b2b0-2f4ce830ec8a | FLAKY         | 20/21    | upstream reliability is '96.80851063829788'. current run reliability is '95.23809523809523'. drift is 1.57042 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66641">build#66641</a></summary>


| test_class                                        | test_method                                       | test_arguments                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                   |
|:--------------------------------------------------|:--------------------------------------------------|:-------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------|
| cluster_metadata_uploader_fixture                 | test_upload_in_term                               |                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/66641#01972299-6205-4368-bde3-a6fa0b35451a | FAIL          | 0/1      |                                                                                                                                                          |
| CloudRetentionTest                                | test_cloud_retention                              | {"cloud_storage_type": 2, "max_consume_rate_mb": null} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66641#019722f7-67b1-41ce-a416-5119f9a3265e | FLAKY         | 20/21    | upstream reliability is '97.5'. current run reliability is '95.23809523809523'. drift is 2.2619 and the allowed drift is set to 50. The test should PASS |
| src/v/cloud_topics/throttler/tests/throttler_test | src/v/cloud_topics/throttler/tests/throttler_test |                                                        | unit        | https://buildkite.com/redpanda/redpanda/builds/66641#01972299-6205-4368-bde3-a6fa0b35451a | FAIL          | 0/1      |                                                                                                                                                          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67106">build#67106</a></summary>


| test_class            | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| raft_fixture          | leadership_transfer_delay            |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/67106#01975a83-5f3a-463f-95da-0edb61667558 | FAIL          | 0/1      |                                                                                                                                                                        |
| DataMigrationsApiTest | test_creating_and_listing_migrations |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67106#01975aea-6161-4dd4-9607-5a3c8f7a1f3f | FLAKY         | 20/21    | upstream reliability is '99.63503649635037'. current run reliability is '95.23809523809523'. drift is 4.39694 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

oleiman committed:
treewide: remove record_batch_reader::storage_t

Removed the variant in the previous commit, so storage_t was a straight 1:1
alias for record_batch_reader::data_t.

oleiman commented:
/ci-repeat
release
skip-redpanda-build
skip-units

oleiman commented:
/ci-repeat 1

dotnwat commented on src/v/model/record_batch_reader.h:
```diff
@@ -47,11 +47,6 @@ concept ReferenceBatchReaderConsumer = requires(
 class record_batch_reader final {
 public:
     using data_t = chunked_circular_buffer<model::record_batch>;
-    struct foreign_data_t {
-        ss::foreign_ptr<std::unique_ptr<data_t>> buffer;
-        size_t index{0};
-    };
```
are there details some where about why this is no longer needed?

oleiman replied:
https://redpandadata.slack.com/archives/C07HD9U0EHL/p1747768459965679?thread_ts=1747766165.349299&cid=C07HD9U0EHL

ballard26 replied:
`foreign_data_t` is no longer needed in general as sharing/reading an `iobuf` was made thread-safe in https://github.com/redpanda-data/seastar/pull/149

oleiman commented:
/ci-repeat 1

dotnwat commented on src/v/model/record_batch_reader.cc:
```diff
@@ -25,8 +25,6 @@
-using storage_t = record_batch_reader::storage_t;
 
 /// \brief wraps a reader into a foreign_ptr<unique_ptr>
 record_batch_reader make_foreign_record_batch_reader(record_batch_reader&& r) {
```
i think what's missing from the pr is maybe a bit of history. when did we stop creating instances of foreign_data_t, and what's the relationship of that to make_foreign_record_Batch_reader

oleiman replied:
yeah. I don't have much historical context. maybe @ballard26 can provide some. presumably there was a previous PR that removed all such usage. happy to chase that down when I have a few spare cycles.

ballard26 replied:
https://github.com/redpanda-data/redpanda/pull/24047 is where the uses of `foreign_data_t` were removed.

oleiman replied:
ah, the cross shard iobuf exercise. makes sense, thanks

dotnwat replied:
thanks @ballard26. @oleiman can we add a reference to that and a little description of the situation to the commit?

oleiman replied:
yeah absolutely. will do this evening.

oleiman replied:
done. @ballard26  can you spot check my description in the PR comment? not sure i've quite nailed it to the wall.
